tests/unit/basket/test_offers.py

```diff
@@ -0,0 +1,159 @@
+from decimal import Decimal as D, ROUND_HALF_UP
+
+from django.test import TestCase, RequestFactory
+from django.urls import reverse
+from django.conf import settings
+
+from oscar.core.loading import get_model, get_class, get_classes
+from oscar.test.factories import UserFactory
+
+Basket = get_model("basket", "Basket")
+Product = get_model("catalogue", "Product")
+factory = RequestFactory()
+Applicator = get_class("offer.applicator", "Applicator")
+Selector, UK = get_classes("partner.strategy", ["Selector", "UK"])
+
+
+class UKSelector(Selector):
+    def strategy(self, request=None, user=None, **kwargs):
+        return UK(request)
+
+
+def money(amount):
+    return amount.quantize(D("0.01"), ROUND_HALF_UP)
+
+
+def get_user_basket(user, request):
+    editable_baskets = Basket.objects.filter(status__in=["Open", "Saved"])
+    basket, __ = editable_baskets.get_or_create(owner=user)
+    basket.strategy = UKSelector().strategy(request=request, user=user)
+    basket.reset_offer_applications()
+    if not basket.is_empty:
+        Applicator().apply(basket, user, request)
+    request.session[settings.OSCAR_BASKET_COOKIE_OPEN] = basket.pk
+    request.session.save()
+    return basket
+
+
+class OfferTest(TestCase):
+    fixtures = ["catalogue", "offer"]
```

Comment:
I suppose we can just use factories for catalogue models and generate necessary offers in the test utils instead of using fixture, which would need to be updated within the models changes.

Reply:
Hi @sasha0, thanks for your feedback, understandable considering the way the rest of oscar's tests are implemented. However,  I think updating the fixtures is much, much safer because they are actually made with the user interface, which means they are a real representation of data users produce. It would be much, much more difficult to update the factories to reflect real offers, especially after changes to the offer machinery (models included). For example there is a field called 'exclusive' which was recently added. Without understanding what that field is for, factory boy would just put some random value or the default in there. This could lead to testing of non real world scenarios. I know fixtures are frowned upon by many people, but they are explicit and they are explicitly updated when models change and I think that is a good thing instead of a burden, See zen of python.